jax_distribution/systems/ff_mappo_rware.py

```diff
@@ -0,0 +1,508 @@
+                    loss_actor = loss_actor.mean()
+                    entropy = actor_policy.entropy().mean()
+
+                    total_loss = (
+                        loss_actor + config["vf_coef"] * value_loss - config["ent_coef"] * entropy
+                    )
+                    return total_loss, (value_loss, loss_actor, entropy)
+
+                grad_fn = jax.value_and_grad(_loss_fn, has_aux=True)
+                loss_info, grads = grad_fn(params, opt_state, traj_batch, advantages, targets)
+
+                # Compute the parallel mean (pmean) over the batch.
+                # This calculation is inspired by the Anakin architecture demo notebook.
+                # available at https://tinyurl.com/26tdzs5x
+                # This pmean could be a regular mean as the batch axis is on the same device.
+                grads, loss_info = jax.lax.pmean((grads, loss_info), axis_name="batch")
+                # pmean over devices.
+                grads, loss_info = jax.lax.pmean((grads, loss_info), axis_name="device")
+
+                updates, new_opt_state = update_fn(grads, opt_state)
+                new_params = optax.apply_updates(params, updates)
+
+                return (new_params, new_opt_state), loss_info
+
+            params, opt_state, traj_batch, advantages, targets, rng = update_state
+            rng, shuffle_rng = jax.random.split(rng)
+
+            # SHUFFLE MINIBATCHES
+            batch_size = config["rollout_length"] * config["num_envs"]
+            permutation = jax.random.permutation(shuffle_rng, batch_size)
+            batch = (traj_batch, advantages, targets)
+            batch = jax.tree_util.tree_map(lambda x: merge_leading_dims(x, 2), batch)
+            shuffled_batch = jax.tree_util.tree_map(
+                lambda x: jnp.take(x, permutation, axis=0), batch
+            )
+            minibatches = jax.tree_util.tree_map(
+                lambda x: jnp.reshape(x, [config["num_minibatches"], -1] + list(x.shape[1:])),
+                shuffled_batch,
+            )
+
+            # UPDATE MINIBATCHES
+            (params, opt_state), loss_info = jax.lax.scan(
+                _update_minibatch, (params, opt_state), minibatches
+            )
+
+            update_state = (params, opt_state, traj_batch, advantages, targets, rng)
+            return update_state, loss_info
+
+        update_state = (params, opt_state, traj_batch, advantages, targets, rng)
+
+        # UPDATE EPOCHS
+        update_state, loss_info = jax.lax.scan(
+            _update_epoch, update_state, None, config["ppo_epochs"]
+        )
+
+        params, opt_state, traj_batch, advantages, targets, rng = update_state
+        learner_state = LearnerState(params, opt_state, rng, env_state, last_timestep)
+        metric = traj_batch.info
+        return learner_state, (metric, loss_info)
+
+    def learner_fn(learner_state: LearnerState) -> ExperimentOutput:
```

Comment:
Nice, thought you might have to change these types, good to see that they're just the same for this case :fire: